protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCompression.java

```diff
@@ -1,88 +1,73 @@
+import org.geysermc.mcprotocollib.network.compression.PacketCompression;
 
 import java.util.List;
-import java.util.zip.Deflater;
-import java.util.zip.Inflater;
 
-public class TcpPacketCompression extends ByteToMessageCodec<ByteBuf> {
-    private static final int MAX_UNCOMPRESSED_SIZE = 8388608;
+public class TcpPacketCompression extends MessageToMessageCodec<ByteBuf, ByteBuf> {
+    private static final int MAX_UNCOMPRESSED_SIZE = 8 * 1024 * 1024; // 8MiB
 
     private final Session session;
-    private final Deflater deflater = new Deflater();
-    private final Inflater inflater = new Inflater();
-    private final byte[] buf = new byte[8192];
+    private final PacketCompression compression;
     private final boolean validateDecompression;
 
-    public TcpPacketCompression(Session session, boolean validateDecompression) {
+    public TcpPacketCompression(Session session, PacketCompression compression, boolean validateDecompression) {
         this.session = session;
+        this.compression = compression;
         this.validateDecompression = validateDecompression;
     }
 
     @Override
-    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
-        super.handlerRemoved(ctx);
-
-        this.deflater.end();
-        this.inflater.end();
+    public void handlerRemoved(ChannelHandlerContext ctx) {
+        this.compression.close();
     }
 
     @Override
-    public void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) {
-        int readable = in.readableBytes();
-        if (readable > MAX_UNCOMPRESSED_SIZE) {
-            throw new EncoderException("Packet too big: size of " + readable + " is larger than the protocol maximum of " + MAX_UNCOMPRESSED_SIZE + ".");
-        }
-        if (readable < this.session.getCompressionThreshold()) {
-            this.session.getCodecHelper().writeVarInt(out, 0);
-            out.writeBytes(in);
+    public void encode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) {
+        int uncompressed = msg.readableBytes();
+        ByteBuf outBuf = ctx.alloc().directBuffer(uncompressed);
+        if (uncompressed < this.session.getCompressionThreshold()) {
+            // Under the threshold, there is nothing to do.
+            this.session.getCodecHelper().writeVarInt(outBuf, 0);
+            outBuf.writeBytes(msg);
         } else {
-            byte[] bytes = new byte[readable];
-            in.readBytes(bytes);
-            this.session.getCodecHelper().writeVarInt(out, bytes.length);
-            this.deflater.setInput(bytes, 0, readable);
-            this.deflater.finish();
-            while (!this.deflater.finished()) {
-                int length = this.deflater.deflate(this.buf);
-                out.writeBytes(this.buf, 0, length);
-            }
-
-            this.deflater.reset();
+            this.session.getCodecHelper().writeVarInt(outBuf, uncompressed);
+            compression.deflate(msg, outBuf);
```

Comment:
we don't want to check for the max uncompressed size anymore?

Reply:
We do using         if (uncompressed < this.session.getCompressionThreshold()) { 
Or is there smth I'm missing?

Reply:
Oh I see now.